PED - UG improvements (#203)
AY2122S1-CS2103T-W08-2/tp: docs/UserGuide.md

If you can type fast, Source Control can help you track your students' performance faster than traditional GUI apps!

This guide takes you through all the _latest features_ of Source Control. If you are a new user, this guide provides you
all the basic knowledge to [get started](#quick-start) with Source Control. If you are an experienced user,
feel free to skip to the [command summary](#command-summary) and experiment with our more [advanced features](#setting-customised-aliases-for-commands--alias)!

--------------------------------------------------------------------------------------------------------------------

## Table of Contents

* [Overview](#)
* [Table of Contents](#table-of-contents)
* [Quick Start](#quick-start)
* Features
  * [Notes on command format](#features)

>   * Adding data

Reviewer: I think here we are capitalizing all section titles for 1st and 2nd levels, right? (By 1st and 2nd levels I mean `Features` is at 1st level and so on)

In that case, should `Note on command format` and `Adding data` be capitalized too, esp. for the latter one since `Searching Data` and so on are also capitalized?

```suggestion
  * [Notes on Command Format](#features)
  * Adding Data
```

    * [Adding a student : `add student`](#adding-a-student--addstudent)
    * [Creating a new group : `add group`](#creating-a-new-group--addgroup)
    * [Adding a student into a group : `add alloc`](#adding-a-student-into-a-group--addalloc)
    * [Creating a new assessment : `add assessment`](#creating-a-new-assessment--addassessment)

>     * [Adding a score : `add score`](#adding-a-score--addscore)

Reviewer: I think you forget to update the command keywords into one-word keywords 😅

```suggestion
    * [Adding a student : `addstudent`](#adding-a-student--addstudent)
    * [Creating a new group : `addgroup`](#creating-a-new-group--addgroup)
    * [Adding a student into a group : `addalloc`](#adding-a-student-into-a-group--addalloc)
    * [Creating a new assessment : `addassessment`](#creating-a-new-assessment--addassessment)
    * [Adding a score : `addscore`](#adding-a-score--addscore)
```

  * Searching Data
    * [Listing all students : `list`](#listing-all-students--list)
    * [Searching for students : `search`](#searching-for-students--search)
  * Analysing Data
    * [Showing assessment result analysis : `show`](#showing-assessment-result-analysis--show)
  * Editing Data
    * [Editing a student : `edit`](#editing-a-student--edit)
    * [Deleting a student : `delete`](#deleting-a-student--delete)

>   * Import and Export

Reviewer: Shouldn't it be `Import and Export Data` to be consistent with other headers?

```suggestion
  * Import and Export Data
```

    * [Importing data : `import`](#importing-data--import)
    * [Exporting data : `export`](#exporting-data--export)

>   * General features

Reviewer: Same here for the capitalization, shouldn't it be `General Features` instead?

```suggestion
  * General Features
```

    * [Resetting all data : `clear`](#resetting-all-data--clear)

>     * [Closing the app : `exit`](#closing-the-app--exit)

Reviewer: Should it be `Clearing all data` and `Exiting the app` to match with the command keywords? 😅

```suggestion
    * [Clearing all data : `clear`](#resetting-all-data--clear)
    * [Exiting the app : `exit`](#closing-the-app--exit)
```

>     * [Viewing Help : `help`](#viewing-help--help)

Reviewer: Same here for the capitalization, shouldn't it be `Viewing help` instead (since it's at 3rd level)?

```suggestion
    * [Viewing help : `help`](#viewing-help--help)
```

    * [Accessing command history](#accessing-command-history)
  * Advanced Features

>     * [Setting customised aliases for commands: `alias`](#setting-customised-aliases-for-commands--alias)

Reviewer: I think it should be "customized" since we have been using the American style throughout the UG? (Maybe no need to change the bookmark `#` haha)

```suggestion
    * [Setting customized aliases for commands: `alias`](#setting-customised-aliases-for-commands--alias)
```

Author: ive searched through the document everything is written as "se" ;-; lets use British spelling please

Reviewer: Ohh okay! I thought we were using American style since I notice this first haha (but it seems like it is the only one...), so yea may I trouble u to change it from "optimized" to "optimised" >;<

![image](https://user-images.githubusercontent.com/63232300/139900927-95633ff6-46a2-486d-9455-4bebf6086dbf.png)

    * [Editing the data file](#editing-the-data-file)
* [FAQ](#faq)
* [Glossary](#glossary)
* [Command Summary](#command-summary)

<br>

--------------------------------------------------------------------------------------------------------------------

## Quick Start

1. Ensure you have Java `11` or above installed in your Computer.

1. Download the latest `sourceControl.jar` from [here](https://github.com/AY2122S1-CS2103T-W08-2/tp/releases).

1. Copy the file to the folder you want to use as the _home folder_ for Source Control.

1. Double-click the file to start the app. The GUI shown in the image below should appear in a few seconds. Note how the app contains some sample data.<br>
   ![Ui](images/StartingUi.png)

> 1. Type the command in the command box and press Enter to execute it. e.g. typing **`help`** and pressing Enter will open the help window.<br>

Reviewer: I think "e.g." needs to be in uppercase, or there should be a comma behind "it"

```suggestion
1. Type the command in the command box and press Enter to execute it, e.g. typing **`help`** and pressing Enter will open the help window.<br>
```

Author: ive changed all similar occurrences to upper case E.g.

> * `<angled brackets>`: <br>

Reviewer: Hmm just notice this now, shouldn't it be just "angle"? Or maybe "diamond" as this sounds more intuitive to me. What do you think?

![image](https://user-images.githubusercontent.com/63232300/139869443-e7fd9724-c4d9-4e8e-b99a-c97a7f40d581.png)

```suggestion
* `<diamond brackets>`: <br>
```

Ps: some changes don't belong to your PR but I would really appreciate if you could help me change them >;<

Author: wow ur right

> [Return to table of contents](#table-of-contents)

Reviewer: Hmmm I think "table of contents" should be in uppercase as we capitalize it in the beginning, i.e. the whole sentence will be `Return to Table of Contents". Same for other places. What do you think?

```suggestion
[Return to Table of Contents](#table-of-contents)
```

* There should not be an existing student with the same NUSNET ID. If there is, the student to be added is considered invalid.

Examples:
* `addstudent -n Jonas Chow -i E0123456` 
  * adds the student Jonas Chow with the given NUSNET ID.
* `addstudent -n Jonas Chow -i E0123456 -g T01A -g R01A` 

>   * adds the student Jonas Chow and allocate him into groups `T01A` and `R01A`.
> * `addstudent -n Jonas Chow -i E0123456 -t beginner` 
>   * adds the student Jonas Chow and tag him with `beginner`.

Reviewer: Shouldn't "allocate" and "tag" be "allocates" and "tags" to be consistent with "adds"?

```suggestion
  * adds the student Jonas Chow and allocates him into groups `T01A` and `R01A`.
* `addstudent -n Jonas Chow -i E0123456 -t beginner` 
  * adds the student Jonas Chow and tags him with `beginner`.
```

Format: `addgroup -g <group_name> [(-n <student_name> | -i <student_id>)]...`

> * Creates a new group with the given group name. Group should not already exist.

Reviewer: Hm should "group" be "the group" instead?

```suggestion
* Creates a new group with the given group name. The group should not already exist.
```

* Group name can be any number of alphanumeric words.
* Students can be identified by their name or NUSNET ID.
Name is case-sensitive and only allows exact match e.g. `Jonas` will not match `Jonas Chow`.
* If multiple students have the same name, NUSNET ID needs to be used to identify them. 

<div markdown="block" class="alert alert-info">
**:information_source: Notes:**<br>

If duplicated students are found in the list of names and NUSNET IDs input, and error would be thrown to inform you of

> the clash, and the group would not be created. 

Reviewer: Should it be "will be thrown" and "will not be created" instead?

```suggestion
If duplicated students are found in the list of names and NUSNET IDs input, and an error will be thrown to inform you of
the clash, and the group will not be created. 
```

* If multiple students have the same name, NUSNET ID needs to be used to identify them.

Examples:
* `addalloc -g T01A -n Zhiying` 

>   * adds `Zhiying` into the tutorial group `T01A`.
> * `addalloc -g T02A -i E0123456` 
>   * adds student with NUSNET ID `E0123456` into tutorial group `T02A`.

Reviewer: Maybe only "tutorial group T01A" on Line `199` to be consistent with Line `201`

```suggestion
  * adds `Zhiying` into tutorial group `T01A`.
* `addalloc -g T02A -i E0123456` 
  * adds student with NUSNET ID `E0123456` into tutorial group `T02A`.
```

* If multiple students have the same name, NUSNET ID needs to be used to identify them.

Examples:
* `addscore -a Midterm -n Van Nhi -s 95` 

>   * records score for `Van Nhi` in `Midterm` to be `95` percent.

Reviewer: Hm should this be `95.00` since we record `95.00` into the system instead of just `95`? Just my opinion what do you think?

```suggestion
  * records score for `Van Nhi` in `Midterm` to be `95.00` percent.
```

Format: `show (<index> | -n <student_name> | -i <student_id> | -g <group_name> | -a <assessment_name> ) [-f <export_file_path>]`

* Using `<index>`, `-n <student_name>` or `-i <student_id>` displays the information of the particular student's performance in all his graded assessments.

>   * Student's score in each assessment, as well as the cohort mean and median score, will be shown as a line graph. 
>   * The `<index>` refers to the index number shown in the displayed student list. The `<index>` must be a positive integer 1, 2, 3, …
>   * If multiple students have the same name, NUSNET ID needs to be used to identify them.
> * Using `-g <group_name>` displays the information of the group's performance in all their graded assessments.
>   * Group's mean score in each assessment, as well as the cohort mean and median score, will be shown as a line graph.
> * Using `-a <assessment_name>` displays the information of the cohort's performance in the particular assessment.
>   * The cohort's score distribution will be shown as a histogram. 

Reviewer: Maybe need slight changes as follows to be consistent

```suggestion
  * The student's score in each assessment, as well as the cohort mean and median score, will be shown as a line graph. 
  * The `<index>` refers to the index number shown in the displayed student list. The `<index>` must be a positive integer 1, 2, 3, …
  * If multiple students have the same name, NUSNET ID needs to be used to identify them.
* Using `-g <group_name>` displays the information of the group's performance in all their graded assessments.
  * The group's mean score in each assessment, as well as the cohort mean and median score, will be shown as a line graph.
* Using `-a <assessment_name>` displays the information of the cohort's performance in the particular assessment.
  * The cohort's score distribution will be shown as a histogram. 
```

![CsvFormatExample](images/CsvFormatExample.png)

The command to import this file would be `import -f student_data.csv -g 2 -a 2 -t 2`. (There are 2 group columns, 2 assessments, and 2 tag columns.)

Examples:
* `import -f /home/prof/CS1101S/student_data.csv -g 2 -a 10 -t 1` 

>   * imports from the absolute path and database has 2 group columns, 10 assessments, and 1 tag column.
> * `import -f student_data.csv -g 5 -a 20` 
>   * imports from the relative path, and database has 5 group columns and 20 assessments.

Reviewer: How about this?

```suggestion
  * imports from the absolute path, and the database has 2 group columns, 10 assessment columns, and 1 tag column.
* `import -f student_data.csv -g 5 -a 20` 
  * imports from the relative path, and the database has 5 group columns and 20 assessment columns.
```

Author: actually importing assessments works differently from group. i'm trying to reinforce that the group columns =! number of groups. using `assessment columns` works too, if u think this format is too confusing

Reviewer: Oh I see. I'm just afraid the ones who will be reviewing our UGDG may spot this. In that case, as you say they work differently so we can explain it later, but hmm if `assessment columns` is not wrong then I think we should make them consistent (since the reviewers won't know that they work differently)

Author: ok sure

* The file path can be either the absolute path or the relative path.
* The relative path will be relative to the folder your jar file is located.
* If the file path is not specified, it will be saved to `sourceControl.csv`.
* The format of the CSV file saved matches exactly the CSV format used by the `import` command.

<div markdown="block" class="alert alert-primary">

:bulb: **Tips:**<br>

* The exported CSV file can be imported again to restore the state of the application when the data was exported

</div>

Examples:
* `export -f student_data.csv` 

>   * exports database as file `student_data.csv`.

Reviewer: Maybe specify the file location where it will be exported to too? Same for `edit` below

[Return to table of contents](#table-of-contents)

### Closing the app : `exit`

Exits the application.

Format: `exit`

[Return to table of contents](#table-of-contents)

### Viewing help : `help`

Shows a message explaining how to access the help page.

![help message](images/helpMessage.png)

Format: `help`

[Return to table of contents](#table-of-contents)

### Accessing command history

> Using the up and down arrow keys on the keyboard retrieves past command lines entered. 

Reviewer: Should it be `Uses...` instead to be consistent with other sections?

```suggestion
Uses the up and down arrow keys on the keyboard retrieves past command lines entered. 
```

Author: hmm compared to:
"Retrieves past command lines entered using the up and down arrow keys. "
i'm more inclined towards the original
"Using the up and down arrow keys on the keyboard retrieves past command lines entered. "

Reviewer: Hmm how about "Scrolls through the command history using the up and down arrow keys"? I do agree that the original sounds smoother, but since the descriptions of other sections always begin with a verb, I think it would be nice if we can follow that convention

![image](https://user-images.githubusercontent.com/63232300/139905267-be561494-90c9-44c4-9983-5855bc95376b.png)

* `<existing_command>` refers to any command keywords e.g. `search`, `addalloc`, `addstudent`.
* The `<alias>` can only be a single alphanumeric word.
* Multiple aliases can be set for each command.
* Default and existing aliases can still be used after new alias is added. e.g. after `alias -c addstudent -as student`, both `student` and `addstudent` can be used.
* The `<existing_command>` can also take any existing alias as an input. The new alias will be added as an alternative to the default command.
* If you find that you have too many aliases, you can remove an alias by using `alias -c <alias> -as <same_alias>`.
* The aliases created will persist between Source Control sessions, as long as the application is in the same directory as it was previously. 

Examples:
* `alias -c addstudent -as example` 
  * adds a new alias to `addstudent` command.
  * `example -n Zhiying -i E1234567` adds student `Zhiying` to database.
* `alias -c addgroup -as example` 

>   * replaces the mapping of the alias `example` to the `addgroup` command. i.e. `example` no longer function as `addstudent`.
>   * `example -g T02A` will create a new group `T02A`.
> * `alias -c example -as example2` 
>   * adds the alias `example2` to the command that `example` currently maps to. i.e. `example2` functions as `addgroup`.

Reviewer: Maybe there needs an extra comma

```suggestion
  * replaces the mapping of the alias `example` to the `addgroup` command, i.e. `example` no longer function as `addstudent`.
  * `example -g T02A` will create a new group `T02A`.
* `alias -c example -as example2` 
  * adds the alias `example2` to the command that `example` currently maps to, i.e. `example2` functions as `addgroup`.
```

* `<existing_command>` refers to any command keywords e.g. `search`, `addalloc`, `addstudent`.
* The `<alias>` can only be a single alphanumeric word.
* Multiple aliases can be set for each command.
* Default and existing aliases can still be used after new alias is added. E.g. after `alias -c addstudent -as student`, both `student` and `addstudent` can be used.
* The `<existing_command>` can also take any existing alias as an input. The new alias will be added as an alternative to the default command.
* If you find that you have too many aliases, you can remove an alias by using `alias -c <alias> -as <same_alias>`.
* The aliases created will persist between Source Control sessions, as long as the application is in the same directory as it was previously. 

Examples:
* `alias -c addstudent -as example` 
  * adds a new alias to `addstudent` command.

>   * `example -n Zhiying -i E1234567` adds student `Zhiying` to database.

Reviewer: Just one v v small thing 🙏🏻 other than this I think it is okay to merge! So I think here it should be "will add student..." instead to match the "pattern" of the other two examples below.

```suggestion
  * `example -n Zhiying -i E1234567` will add student `Zhiying` to the database.
```
